services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java

```diff
@@ -209,6 +210,22 @@ private boolean validateUser(AuthenticationFlowContext context, UserModel user,
     }
 
     public boolean validatePassword(AuthenticationFlowContext context, UserModel user, MultivaluedMap<String, String> inputData, boolean clearUser) {
+        // If we have validated the password before, but it didn't match the current password policy, allow to continue to
+        // update the password or cancel. If we don't check whether the password was validated, users could skip the authenticator simply
+        // by providing continueUpdate in their form data.
+        if (AuthenticatorUtil.isPasswordValidated(context.getAuthenticationSession())){
+            if(inputData.containsKey("continueToUpdate")) {
+                // Set a required action on the authentication session instead of on the user, as the password policy could be
+                // changed again such that the password would be compliant again. Setting a update action on the user could
+                // make it redundant, if their password is compliant again after a change in the policy.
+                context.getAuthenticationSession().addRequiredAction(UserModel.RequiredAction.UPDATE_PASSWORD);
+                return true;
+            }
+            if(inputData.containsKey("cancelUpdate")) {
+                return badPasswordHandler(context, user, clearUser, false);
```

Comment:
Is this proper use? The password itself was ok, wasn't it?

This would also send a password error mail, doesn't it?. Is this intended?

Reply:
You're right, I tried to reuse the existing functionality, but we I agree that might not want the invalid credentials event (which might trigger an email, depending on the settings of course).